internal/pkg/bully/server/bully.go

```diff
@@ -0,0 +1,239 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package bully
+
+import (
+	"context"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+	"github.com/talos-systems/talos/internal/pkg/bully/proto"
+	"google.golang.org/grpc"
+)
+
+// Address represents the address of a bully.
+type Address = string
+
+// Pid represents a process pid.
+type Pid = uint32
+
+// Processes represents the set of processes that form the cluster.
+type Processes struct {
+	*sync.RWMutex
+
+	internal map[Pid]*Process
+}
+
+// Process represents a bully process.
+type Process struct {
+	*proto.Process
+
+	conn   *grpc.ClientConn
+	client proto.BullyClient
+}
+
+// Bully represents the bully algorithm.
+type Bully struct {
+	Processes *Processes
+
+	broadcast   chan *proto.MessageWithPid
+	addrs       []Address
+	self        *Process
+	coordinator *Process
+}
+
+// NewBullyServer initializes and returns Bully.
+func NewBullyServer(pid Pid, addr Address, addrs ...Address) *Bully {
+	procs := &Processes{
+		RWMutex:  &sync.RWMutex{},
+		internal: map[Pid]*Process{},
+	}
+
+	self := &Process{
+		Process: &proto.Process{
+			Address: addr,
+			Pid:     &proto.PidValue{Value: pid},
+		},
+	}
+
+	bully := &Bully{
+		Processes: procs,
+		broadcast: make(chan *proto.MessageWithPid, len(addrs)*3),
+		addrs:     addrs,
+		self:      self,
+	}
+
+	return bully
+}
+
+// Start registers and starts the gRPC server.
+func (b *Bully) Start() error {
+	l, err := net.Listen("unix", b.self.Process.Address)
+	if err != nil {
+		return err
+	}
+
+	s := grpc.NewServer()
+	proto.RegisterBullyServer(s, b)
+
+	return s.Serve(l)
+}
+
+// Join initializes a bully's set of processes.
+func (b *Bully) Join() error {
+	for _, addr := range b.addrs {
+		sock := "unix://" + addr
+		conn, err := grpc.Dial(sock, grpc.WithInsecure())
+		if err != nil {
+			return err
+		}
+
+		c := proto.NewBullyClient(conn)
+		reply, err := c.Pid(context.Background(), &empty.Empty{})
+		if err != nil {
```

Comment:
this triggers a question - if join fails when any other process is not alive, then what's the purpose of the bully algorithm here? if all the nodes are alive, we don't need to run elections - we can pick the one with the highest PID.

I believe the whole idea of bully algorithm is to skip failed nodes.

in our case, as I understand it, we run election among nodes, so natural thing for PID might be IP address of the node, but if that are floating IPs, and we know the list of IPs beforehand, we can always say that the master is the process with highest IP.

if we run election only on bootstrap, we don't need to detect failed processes? we just say that master has the highest IP, and every other node waits for the master to complete bootstrap?

am I missing something here?

or in other words, part of this bully algorithm is to figure out PIDs of other processes, while it can't do that if some processes don't respond which contradicts some of the algorithm assumptions - algorithm can't order PIDs if it doesn't have the full picture.

Reply:
> this triggers a question - if join fails when any other process is not alive, then what's the purpose of the bully algorithm here? if all the nodes are alive, we don't need to run elections - we can pick the one with the highest PID.

Ahh good points. I think simple solution here is to not fail hard here then. And just to be clear, "join" here is exclusive of `kubeadm join`.

> in our case, as I understand it, we run election among nodes, so natural thing for PID might be IP address of the node, but if that are floating IPs, and we know the list of IPs beforehand, we can always say that the master is the process with highest IP.
I like the idea of using the IP as the PID. Makes it one less thing to configure here.


>if we run election only on bootstrap, we don't need to detect failed processes? we just say that master has the highest IP, and every other node waits for the master to complete bootstrap?

You are correct. There is another component we see making use of bully, and that is proxyd. The idea behind proxyd is that it will be able to detect failed master nodes and dynamically detach/attach a floating IP to the current leader. This is to provide an external HA control plane. So using bully for bootstrap _is_ a bit much, but we have potentially more use cases for bully.

> or in other words, part of this bully algorithm is to figure out PIDs of other processes, while it can't do that if some processes don't respond which contradicts some of the algorithm assumptions - algorithm can't order PIDs if it doesn't have the full picture.

If I'm understanding the algorithm correctly, each process needs to know the address and IP of each other process. So yes, I think you are right. We need to skip the fetching of the PID. The PIDs should be supplied before hand. Am I understanding this question correctly?